rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+
+If users create iterators or other resources inside a function but don’t expose them as outputs, we will not rebuild them.
+
+
+#### The Unknown of Scheduled Functions
+
+For functions that have been scheduled, it is difficult for the client to know whether they have actually been executed or not when the client detects their corresponding worker failure. Therefore, in addition to inform users of this uncertainty in the case of worker failure, we should do the following to reduce this uncertainty:
+
+
+*   keep the number of scheduled but not executed functions small. This may be difficult to achieve since there is not an easy way for the client to know whether a function is executed or not. The only way is to synchronize the executor. Therefore, as a workaround we will have to periodically synchronize the executor to make sure functions are actually executed, before the client schedules more functions. In the long run, we should get acknowledgement from runtime about how many functions have been executed.
+*   eagerly fetch the outputs of remote functions once the outputs are passed to `strategy.local_result`. In this way, we can know the status of function execution earlier.
+*   recommend users schedule only small functions. Large functions are more expensive to retry.
+
+
+### Evaluation
+
+Historically, `tf.estimator.Estimator` uses a dedicated evaluator that periodically loads from a checkpoint, and performs evaluation with evaluation data. However `tf.keras` typically evaluates in an alternating manner after every epoch of training, and this is also the case with `tf.keras` + `MultiWorkerMirroredStrategy`. 
+
+With `ParameterServerStrategyV2`, we will start with a dedicated** evaluator that runs alongside the training cluster**, **aka “sidecar evaluation”**; in this scheme, training client is required to generate checkpoints periodically, and the evaluator reads the latest checkpoint as it becomes available. The evaluation is asynchronous to the training progress. With this we provide the functionality Estimator has been able to with Keras API, which is important to attract updates from Estimator users to TF 2.0.
+
+With our recommendation, users should create a separate evaluation client that runs the same python binary as the training client. This python binary will contain the if-else clause as it bifurcates into two paths:
+
+
+```python
+if cluster_resolver.task_type == "chief":
+  run_training_loop()
+elif cluster_resolver.task_type == "evaluator":
+  run_evaluation_loop()
+```
+
+
+Evaluation code:
+
+
+```python
+def run_evaluation_loop(...):
+  """Run the example custom evaluation loop."""
+  
+  eval_dataset, model, eval_accuracy = ...
+  checkpoint = tf.train.Checkpoint(model=model)
+
+  @tf.function
+  def eval_fn(eval_dataset):
+    for _ in range(eval_steps):
+      # evaluation pass
+    return eval_accuracy.result()
+
+  while True:
```

Comment:
There's definitely room for non-parameter-server-specific utilities here for side car evaluation. Plenty of users (e.g. object detection) want to use dedicated evaluation jobs with mirroredstrategy, tpustrategy, & others.

Reply:
The common patterns are:
Wait until a checkpoint is present
Evaluate
Check if a new one is present, if not wait a while until it is
Finish evaluating when no new checkpoint has been written for a while

With timeouts, meaningful error messages, etc.

Reply:
Exactly and agreed. We're working on this and will update the design doc accordingly. Thanks for pointing this out.